File: handler/writer.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
MENU_TEMPLATE = "    \"{label}\":\n        jump {target}\n"
SIDE_CHARACTER_TEMPLATE = "image side {role_name} = \"{path}\"\n"


class RpyFileWriter(object):

    @classmethod
    def write_file(cls, output_dir, res, role_name_mapping, role_side_character_mapping):
        output_path = output_dir + "/" + res.label + '.rpy'
        with open(output_path, 'w', encoding='utf-8') as f:
            for k, v in role_name_mapping.items():
                f.write(v.render() + "\n")
            f.write("define narrator_nvl = Character(None, kind=nvl)\n")
            f.write("define narrator_adv = Character(None, kind=adv)\n")
            f.write("define config.voice_filename_format = \"audio/{filename}\"\n")
            for k, v in role_side_character_mapping.items():
                f.write(SIDE_CHARACTER_TEMPLATE.format(role_name=k, path=v))
            f.write("\nlabel {}:\n".format(res.label))
            last_voice = None
            current_menus = []
            for rpy_element in res.data:
                if rpy_element.menu:
                    current_menus.append(rpy_element.menu)
                    continue
                if current_menus:
                    f.write("menu:\n" + "\n".join(
                        [MENU_TEMPLATE.format(label=m.label, target=m.target) for m in current_menus]))
                    current_menus.clear()
                    continue
                if rpy_element.music:
                    f.write(rpy_element.music.render() + '\n')
                if rpy_element.character:
                    for ch in rpy_element.character:
                        f.write(ch.render() + '\n')
                if rpy_element.background:
                    f.write(rpy_element.background.render() + '\n')
                if rpy_element.sound:
                    f.write(rpy_element.sound.render() + '\n')
                if rpy_element.transition:
                    f.write(rpy_element.transition.render() + '\n')
                if rpy_element.voice:
                    f.write(rpy_element.voice.render() + '\n')
                if rpy_element.text:
                    if last_voice and last_voice.sustain:
                        f.write("voice sustain\n")
                    f.write(rpy_element.text.render() + '\n')
                if rpy_element.change_page:
                    f.write(rpy_element.change_page.render() + '\n')
                last_voice = rpy_element.voice
            if current_menus:
                # fix menu在最后一行
                f.write("menu:\n" + "\n".join(
                    [MENU_TEMPLATE.format(label=m.label, target=m.target) for m in current_menus]))

File: packages/core/src/write/writer.ts
// 写出 .rpy（移植自 handler/writer.py）。逐字节对齐，见 docs/01-legacy-system-contract.md §5。
import type { Role, Voice, Menu } from '../model/element'
import type { SheetConvertResult } from '../convert/converter'

const menuTemplate = (label: string, target: string): string =>
  `    "${label}":\n        jump ${target}\n`
const sideCharacterTemplate = (roleName: string, path: string): string =>
  `image side ${roleName} = "${path}"\n`

export function renderFile(
  res: SheetConvertResult,
  roleNameMapping: Map<string, Role>,
  sideCharacters: Map<string, string>,
): string {
  let out = ''
  // 角色定义（跨 sheet 累积的全部角色，每个文件都完整写出）
  for (const role of roleNameMapping.values()) {
    out += role.render() + '\n'
  }
  out += 'define narrator_nvl = Character(None, kind=nvl)\n'
  out += 'define narrator_adv = Character(None, kind=adv)\n'
  out += 'define config.voice_filename_format = "audio/{filename}"\n'
  for (const [pronoun, path] of sideCharacters) {
    out += sideCharacterTemplate(pronoun, path)
  }
  out += `\nlabel ${res.label}:\n`

  let lastVoice: Voice | null = null
  let currentMenus: Menu[] = []

  for (const el of res.data) {
    if (el.menu) {
      currentMenus.push(el.menu)
      continue
    }
    if (currentMenus.length) {
      // 注意旧行为：刷新 menu 块后 continue，会跳过该行其它元素（既有怪癖，勿"修复"）。
      out += 'menu:\n' + currentMenus.map((m) => menuTemplate(m.label, m.target)).join('\n')
      currentMenus = []
      continue
    }
    if (el.music) out += el.music.render() + '\n'
    if (el.background) out += el.background.render() + '\n'
    // 空数组在旧 Python 中为假值会被跳过；这里显式判断 length。
    if (el.character && el.character.length) {
      for (const ch of el.character) out += ch.render() + '\n'
    }
    if (el.sound) out += el.sound.render() + '\n'
    if (el.transition) out += el.transition.render() + '\n'
    if (el.voice) out += el.voice.render() + '\n'
    if (el.text) {
      if (lastVoice && lastVoice.sustain) out += 'voice sustain\n'
      out += el.text.render() + '\n'
    }
    if (el.changePage) out += el.changePage.render() + '\n'
    lastVoice = el.voice
  }

  if (currentMenus.length) {
    // 修复：menu 在最后一行时也要刷新
    out += 'menu:\n' + currentMenus.map((m) => menuTemplate(m.label, m.target)).join('\n')
  }

  return out
}
